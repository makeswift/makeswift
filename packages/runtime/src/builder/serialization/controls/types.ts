import { DeserializedFunction, SerializedFunction } from '../function-serialization'

type AnyFunction = (...args: any[]) => any

export type Serialize<T> = T extends AnyFunction
  ? SerializedFunction<T>
  : T extends Record<string, unknown>
    ? { [K in keyof T]: Serialize<T[K]> }
    : T

export type Deserialize<T> =
  T extends SerializedFunction<infer U>
    ? DeserializedFunction<U>
    : T extends Record<string, unknown>
      ? { [K in keyof T]: Deserialize<T[K]> }
      : T
