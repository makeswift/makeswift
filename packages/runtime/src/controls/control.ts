import {
  copyRichTextData,
  type CopyContext,
  type Data,
  type MergeContext,
} from '@makeswift/controls'

import { Types as PropControllerTypes } from '@makeswift/prop-controllers'

import { Descriptor, isLegacyDescriptor } from '../prop-controllers/descriptors'
import { copy as propControllerCopy } from '../prop-controllers/copy'
import { DELETED_PROP_CONTROLLER_TYPES } from '../prop-controllers/deleted'

export function copy(definition: Descriptor, value: any, context: CopyContext) {
  if (!isLegacyDescriptor(definition)) {
    return definition.copyData(value, context)
  }

  switch (definition.type) {
    case PropControllerTypes.Backgrounds:
    case PropControllerTypes.Grid:
    case PropControllerTypes.NavigationLinks:
    case PropControllerTypes.Link:
    case PropControllerTypes.Shadows:
    case PropControllerTypes.Image:
    case PropControllerTypes.Images:
    case PropControllerTypes.ResponsiveColor:
    case PropControllerTypes.TableFormFields:
    case PropControllerTypes.Table:
    case PropControllerTypes.Border:
    case PropControllerTypes.ElementID:
      return propControllerCopy(definition, value, context)

    case DELETED_PROP_CONTROLLER_TYPES.RichText:
      return copyRichTextData(value, context)

    default:
      return value
  }
}

export function merge(definition: Descriptor, a: Data, b: Data = a, context: MergeContext): Data {
  if (!isLegacyDescriptor(definition)) {
    return definition.mergeData(a, b, context)
  }

  switch (definition.type) {
    default:
      return b
  }
}

export function getTranslatableData(definition: Descriptor, data: Data): Data {
  if (!isLegacyDescriptor(definition)) {
    return definition.getTranslatableData(data)
  }

  switch (definition.type) {
    case PropControllerTypes.TextInput:
    case PropControllerTypes.TextArea:
      return data

    default:
      return null
  }
}
