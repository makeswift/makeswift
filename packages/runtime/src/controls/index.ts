export * from './control'

export {
  ControlDefinition,
  Checkbox,
  CheckboxDefinition,
  Color,
  ColorDefinition,
  Combobox,
  ComboboxDefinition,
  Font,
  FontDefinition,
  Group,
  GroupDefinition,
  GroupControl,
  IconRadioGroupDefinition,
  type IconRadioGroupIcon,
  unstable_IconRadioGroup,
  Image,
  ImageDefinition,
  List,
  ListDefinition,
  ListControl,
  Number,
  NumberDefinition,
  Select,
  SelectDefinition,
  Shape,
  ShapeDefinition,
  ShapeControl,
  Style,
  StyleDefinition,
  type StyleProperty,
  StyleControl,
  TextInput,
  TextInputDefinition,
  TextArea,
  TextAreaDefinition,
  unstable_Typography,
  unstable_TypographyDefinition,
} from '@makeswift/controls'

export { Link, LinkDefinition } from './link'
export { RichTextV1Definition, RichTextV1Control } from './rich-text'
export { RichText, RichTextV2Definition, RichTextV2Control } from './rich-text-v2'
export { Slot, SlotDefinition, SlotControl } from './slot'
export { unstable_StyleV2, StyleV2Definition, StyleV2Control } from './style-v2/style-v2'
