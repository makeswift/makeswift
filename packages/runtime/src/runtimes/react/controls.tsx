import { useRef, ReactNode } from 'react'

import { ControlDefinition } from '@makeswift/controls'

import { partitionRecord } from '../../utils/partition'

import * as ReactPage from '../../state/react-page'
import { isLegacyDescriptor, LegacyDescriptor } from '../../prop-controllers/descriptors'

import { useStore } from './hooks/use-store'

import { resolveLegacyDescriptorProp } from './legacy-controls'
import { useResolvedProps } from './hooks/use-resolved-props'

type PropsValueProps = {
  element: ReactPage.ElementData
  children(props: Record<string, unknown>): ReactNode
}

function useControlDefs(
  elementType: string,
): readonly [Record<string, LegacyDescriptor>, Record<string, ControlDefinition>] {
  const store = useStore()
  const all = ReactPage.getComponentPropControllerDescriptors(store.getState(), elementType) ?? {}
  return useRef(partitionRecord(all, isLegacyDescriptor)).current
}

export function ResolveProps({ element, children: renderComponent }: PropsValueProps): ReactNode {
  const [legacyDescriptors, definitions] = useControlDefs(element.type)

  const resolvedProps = useResolvedProps(definitions, element.props, element.key)

  const renderFn = Object.entries(legacyDescriptors).reduceRight(
    (renderFn, [propName, descriptor]) =>
      props =>
        resolveLegacyDescriptorProp(descriptor, propName, element.props[propName], props, renderFn),
    renderComponent,
  )

  return renderFn(resolvedProps)
}
