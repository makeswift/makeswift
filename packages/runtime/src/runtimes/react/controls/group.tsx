import {
  GroupControl,
  GroupDefinition,
  type DataType,
  type ResolvedValueType,
} from '@makeswift/controls'

import { ControlValue } from './control'

type GroupControlValueProps = {
  definition: GroupDefinition
  data: DataType<GroupDefinition> | undefined
  children(value: ResolvedValueType<GroupDefinition>): JSX.Element
  control: GroupControl
}

export function GroupControlValue({ definition, data, children, control }: GroupControlValueProps) {
  return Object.entries(definition.config.props).reduceRight(
    (renderFn, [key, controlDefinition]) =>
      shapeControlValue => (
        <ControlValue
          definition={controlDefinition}
          data={data != null ? GroupDefinition.propsData(data)[key] : undefined}
          control={control?.child(key)}
        >
          {value => renderFn({ ...shapeControlValue, [key]: value })}
        </ControlValue>
      ),
    children,
  )({} as ResolvedValueType<GroupDefinition>)
}
