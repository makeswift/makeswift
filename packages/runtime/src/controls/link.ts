import { z } from 'zod'
import { type MouseEvent } from 'react'
import scrollIntoView from 'scroll-into-view-if-needed'

import {
  type DeserializedRecord,
  LinkDefinition as BaseLinkDefinition,
  LinkSchema,
} from '@makeswift/controls'

type DataType = z.infer<typeof LinkDefinition.schema.data>
type ScrollOptions = z.infer<typeof LinkSchema.scrollOptions>
type MouseEventType = MouseEvent<Element>

export function validateElementRef(href: string) {
  try {
    const hash = new URL(`http://www.example.com/${href}`).hash
    return href === hash ? hash : undefined
  } catch (error) {
    return undefined
  }
}

export class LinkDefinition extends BaseLinkDefinition<MouseEventType> {
  static deserialize(data: DeserializedRecord): LinkDefinition {
    if (data.type !== LinkDefinition.type) {
      throw new Error(`Link: expected type ${LinkDefinition.type}, got ${data.type}`)
    }

    const { config } = LinkDefinition.schema.definition.parse(data)
    return Link(config)
  }

  resolveOnClick(
    data: DataType | undefined,
    href: string,
    scrollOptions: ScrollOptions | undefined,
  ) {
    const onClick = (event: MouseEvent<Element>) => {
      if (event.defaultPrevented) return

      if (data && data.type === 'SCROLL_TO_ELEMENT') {
        const hash = validateElementRef(href)
        if (hash == null) {
          console.error(`Scroll-to-element link received invalid href: ${href}`)
          return
        }

        event.preventDefault()
        const view = event.view as unknown as Window

        scrollIntoView(view.document.querySelector(hash)!, {
          behavior: 'smooth',
          block: scrollOptions?.block,
        })

        if (view.location.hash !== hash) view.history.pushState({}, '', hash)
      }
    }

    onClick.$scrollOptions = scrollOptions
    return onClick
  }
}

export function Link(config?: { description?: string, label?: string }): LinkDefinition {
  return new LinkDefinition(config ?? {})
}
