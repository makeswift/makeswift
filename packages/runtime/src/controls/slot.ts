import { ReactNode } from 'react'
import {
  SlotDefinition as BaseSlotDefinition,
  SlotControl,
  StableValue,
  type DeserializedRecord,
  type ResourceResolver,
  type Stylesheet,
  type Resolvable,
  type DataType,
} from '@makeswift/controls'

import { renderSlot } from '../runtimes/react/controls/slot'

type Config = {
  columnCount?: number
}

abstract class BaseDefinition extends BaseSlotDefinition<ReactNode> {}

export class SlotDefinition extends BaseDefinition {
  static deserialize(data: DeserializedRecord): SlotDefinition {
    if (data.type !== SlotDefinition.type) {
      throw new Error(`Slot: expected type ${SlotDefinition.type}, got ${data.type}`)
    }

    const { config } = data
    return Slot(typeof config === 'object' && config != null ? (config as Config) : undefined)
  }

  resolveValue(
    data: DataType<BaseDefinition> | undefined,
    _resolver: ResourceResolver,
    _stylesheet: Stylesheet,
    control?: SlotControl,
  ): Resolvable<ReactNode | undefined> {
    const stableValue = StableValue({
      name: SlotDefinition.type,
      read: () => renderSlot({ data, control: control ?? null, config: this.config as Config }),
    })

    return {
      ...stableValue,
      triggerResolve: async () => {},
    }
  }
}

export function Slot(config?: Config): SlotDefinition {
  return new SlotDefinition(config as unknown as never)
}

export { SlotControl }
