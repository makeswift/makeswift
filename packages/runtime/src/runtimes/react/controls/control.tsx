'use client'

import {
  ControlDefinition,
  type InstanceType,
  type DataType,
  type ResolvedValueType,
} from '@makeswift/controls'
import { ReactNode } from 'react'

import { useResolvedValue } from '../hooks/use-resolved-value'
import { useStylesheetFactory } from '../hooks/use-stylesheet-factory'
import { useCssId } from '../hooks/use-css-id'

type ControlValueProps = {
  definition: ControlDefinition
  data: DataType<ControlDefinition> | undefined
  children(value: ResolvedValueType<ControlDefinition>): ReactNode
  control?: InstanceType<ControlDefinition>
}

export function ControlValue({
  data,
  definition,
  children,
  control,
}: ControlValueProps): ReactNode {
  const stylesheetFactory = useStylesheetFactory()
  const id = `cv-${useCssId()}`

  const value = useResolvedValue(
    data,
    (data, resourceResolver) =>
      definition.resolveValue(data, resourceResolver, stylesheetFactory.get(id), control),
    (definition.config as any)?.defaultValue,
  )

  stylesheetFactory.useDefinedStyles()

  return children(value)
}
