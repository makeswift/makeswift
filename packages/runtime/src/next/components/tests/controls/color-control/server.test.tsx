import { Color, ColorData } from '@makeswift/controls'
import { testPageControlPropRendering } from '../page-control-prop-rendering'
import { swatch, value, cacheData } from './fixtures'

describe.each([
  {
    version: 0,
    toData: (value: ColorData) => value,
  },
  {
    version: 1,
  },
])('color data v$version', ({ toData }) => {
  test.each([
    {
      value,
      swatch,
      defaultValue: 'rgb(255,0,0)',
    },
    {
      value,
      swatch: null,
      defaultValue: 'rgb(255,0,0)',
    },
  ])(`reading swatch $swatch from cache`, async ({ value, defaultValue, swatch }) => {
    await testPageControlPropRendering(Color({ defaultValue }), {
      toData,
      value,
      cacheData: cacheData(swatch),
    })
  })
})
