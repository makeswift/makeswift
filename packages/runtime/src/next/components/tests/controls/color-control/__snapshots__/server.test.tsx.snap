// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`color data v0 reading swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} from cache: component styles 1`] = `[]`;

exports[`color data v0 reading swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} from cache: resolvedValue 1`] = `"rgba(16, 24, 234, 0.5)"`;

exports[`color data v0 reading swatch null from cache: component styles 1`] = `[]`;

exports[`color data v0 reading swatch null from cache: resolvedValue 1`] = `"rgb(255, 0, 0)"`;

exports[`color data v1 reading swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} from cache: component styles 1`] = `[]`;

exports[`color data v1 reading swatch {"__typename": "Swatch", "hue": 238, "lightness": 49, "saturation": 87} from cache: resolvedValue 1`] = `"rgba(16, 24, 234, 0.5)"`;

exports[`color data v1 reading swatch null from cache: component styles 1`] = `[]`;

exports[`color data v1 reading swatch null from cache: resolvedValue 1`] = `"rgb(255, 0, 0)"`;
